src/main/java/gov/nasa/jpl/mbee/mdk/generator/DocumentAndPdfWriter.java

```diff
@@ -0,0 +1,197 @@
+import gov.nasa.jpl.mbee.mdk.model.Document;
+import org.apache.commons.lang.exception.ExceptionUtils;
+import org.apache.fop.apps.*;
+
+import javax.xml.transform.Result;
+import javax.xml.transform.Transformer;
+import javax.xml.transform.TransformerException;
+import javax.xml.transform.TransformerFactory;
+import javax.xml.transform.sax.SAXResult;
+import javax.xml.transform.stream.StreamSource;
+import java.io.*;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.List;
+
+
+/**
+ * runs the generation as a runnable to not stall magicdraw main thread, also
+ * allow user to cancel
+ * <p>
+ * MD Model -> Document(xml) -> PDF
+ *
+ * @author DavidWillard, Miyako Wilson
+ */
+public class DocumentAndPdfWriter implements RunnableWithProgress {
+
+	private final Document dge;
+	private final boolean genNewImage;
+	private final File docbookXslFo;
+	private final File outputFile;
+	private File docbook;
+
+	public DocumentAndPdfWriter(File docbook, File docbookXslFo, File outputFile) {
+		this(null, false, docbook, docbookXslFo, outputFile);
+	}
+
+	public DocumentAndPdfWriter(Document dge, boolean genNewImage, File docbook) {
+		this(dge, genNewImage, docbook, null, null);
+	}
+
+	public DocumentAndPdfWriter(Document dge, boolean genNewImage, File docbook, File docbookXslFo, File outputFile) {
+		this.dge = dge;
+		this.genNewImage = genNewImage;
+		this.docbook = docbook;
+		this.docbookXslFo = docbookXslFo;
+		this.outputFile = outputFile;
+	}
+
+	protected boolean mdModel2Document(ProgressStatus status) {
+
+		//MD Model -> Document
+		try {
+			BufferedWriter writer = new BufferedWriter(new FileWriter(docbook));
+
+			SessionManager.getInstance().createSession(Application.getInstance().getProject(), DocBookOutputVisitor.class.getSimpleName());
+			DocBookOutputVisitor visitor = new DocBookOutputVisitor(false, docbook.getParentFile().getAbsolutePath());
+			dge.accept(visitor);
+			SessionManager.getInstance().closeSession(Application.getInstance().getProject());
+			DBBook book = visitor.getBook();
+			if (book != null) {
+				DBSerializeVisitor v = new DBSerializeVisitor(genNewImage, docbook.getParentFile(), status);
+				book.accept(v);
+				writer.write("<?xml version=\"1.0\" encoding=\"UTF-8\" ?>\n");
+				writer.write(v.getOut());
+			}
+			writer.flush();
+			writer.close();
+			Application.getInstance().getGUILog().log("A DocBook xml file is created as " + docbook.getAbsolutePath());
+			return true;
+		} catch (IOException e) {
+			Application.getInstance().getGUILog().log("[ERROR] Failed to generate a DocBook XML file. " + ExceptionUtils.getStackTrace(e));
+			return false;
+		}
+
+	}
+
+	//similar to mdModel2Document but not create xml in temp directory and remove when the application ended.
+	//no message to Notification window
+	//used by "DocGen->Generate->PDF" action
+	protected boolean mdModel2DocumentAsTemp(ProgressStatus status) {
```

Comment:
This duplication should no longer be necessary once `DocumentAndPdfWriter` is separated and refactored. See other comment.

Reply:
mdModel2DocumentAsTemp is kept  but to create docbook in temp directory and to use DocumentWriter.mdModel2Document.

Reply:
Rename to PdfWriter.